src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2006,3 +2062,444 @@ ShouldInitiateMetadataSync(bool *lockFailure)
 	*lockFailure = false;
 	return shouldSyncMetadata;
 }
+
+
+/*
+ * citus_internal_add_partition_metadata is an internal UDF to
+ * add a row to pg_dist_partition.
+ */
+Datum
+citus_internal_add_partition_metadata(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	char distributionMethod = PG_GETARG_CHAR(1);
+	int colocationId = PG_GETARG_INT32(3);
+	char replicationModel = PG_GETARG_CHAR(4);
+
+	text *distributionColumnText = NULL;
+	char *distributionColumnString = NULL;
+	Var *distributionColumnVar = NULL;
+
+	if (!PG_ARGISNULL(2))
+	{
+		distributionColumnText = PG_GETARG_TEXT_P(2);
+		distributionColumnString = text_to_cstring(distributionColumnText);
+		distributionColumnVar = stringToNode(distributionColumnString);
+	}
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		/* only owner of the table (or superuser) is allowed to add the Citus metadata */
+		EnsureTableOwner(relationId);
+
+		/* this UDF is not allowed allowed for executing as a separate command */
+		EnsureCoordinatorInitiatedOperation();
+
+		if (distributionColumnVar != NULL && distributionMethod == DISTRIBUTE_BY_NONE)
+		{
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("Reference or local tables cannot have "
+								   "distribution columns")));
+		}
+		else if (distributionColumnVar != NULL)
+		{
+			EnsureDistributionKeyExists(relationId, distributionColumnVar);
+		}
+
+		/*
+		 * Even if the table owner is a malicious user and the partition
+		 * metadata is not sane, the user can only affect its own tables.
+		 * Given that the user is owner of the table, we should allow.
+		 */
+		EnsurePartitionMetadataIsSane(distributionMethod, colocationId,
+									  replicationModel);
+	}
+
+	InsertIntoPgDistPartition(relationId, distributionMethod, distributionColumnVar,
+							  colocationId, replicationModel);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * EnsureDistributionKeyExists ensures whether the given distributionColumnVar
+ * exists in the given relationId. If not, the function throws an error.
+ */
+static void
+EnsureDistributionKeyExists(Oid relationId, Var *distributionColumnVar)
+{
+	if (distributionColumnVar == NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Unexpected internal state, distribution "
+							   "column cannot be NULL.")));
+	}
+
+	/*
+	 * Make sure that the column exists on the given table, and print it to prevent any
+	 * optimization skipping the get_attname call.
```

Comment:
Hmm, why do you think so? We are asking PG to give the name of the column? The other alternative is to read pg_attribute but I couldn't see what would be the difference?

Reply:
I mean that we need to log to avoid optimization skipping the call seems really weird.

Reply:
Well, this was a trick Andres suggested for some other place, and we have it in the code: https://github.com/citusdata/citus/blob/master/src/backend/distributed/shared_library_init.c#L408-L418. Because we really suspected that it was optimized by the compiler, and it has been very challenging to find what is going on :) 

 I don't have a strong opinion on switching to read `pg_attribute` to avoid but that seem more code than `get_attname`, so I'll let you make the call. If you prefer so, let me do that

Reply:
We don't need this code anymore, as we pass the column name